cassandane/Cassandane/Cyrus/ALPN.pm

```diff
@@ -253,7 +256,170 @@ sub test_imaps_bad
 
     $self->assert_not_null($e);
     $self->assert_matches($alpn_fail_pattern, $e);
+}
+
+sub do_https_request
+{
+    my ($self, $service, $alpn_map) = @_;
+
+    my $ca_file = abs_path("data/certs/cacert.pem");
+
+    my $client = HTTP::Tiny->new(verify_SSL => 1,
+                                 SSL_options => {
+                                    SSL_ca_file => $ca_file,
+                                    SSL_verifycn_scheme => 'none',
+                                    SSL_alpn_protocols => $alpn_map,
+                                 });
+
+    my $url = sprintf("https://%s:%s@%s:%s/dav/calendars",
+                      'cassandane', 'secret',
+                      $service->host(), $service->port());
+
+    my $content = x('D:propfind', { 'xmlns:D' => 'DAV:' },
+                    x('D:prop',
+                      x('D:current-user-principal')));
+
+    my $response = $client->request('PROPFIND', $url, {
+        headers => {
+            'content-type' => 'application/xml; charset=utf8',
+        },
+        content => "$content", # stringify the XML::Spice::Chunk
+    });
+
+    return $response;
+}
+
+sub test_https_none
+    :want_service_https :needs_component_httpd
+{
+    my ($self) = @_;
+
+    # skip past setup logs
+    $self->{instance}->getsyslog();
+
+    my $https = $self->{instance}->get_service('https');
+    my $alpn_map = undef;
+
+    my $response = $self->do_https_request($https, $alpn_map);
+
+    $self->assert_num_equals(1, $response->{success});
+    $self->assert_alpn_protocol(undef, undef);
+}
+
+sub test_https_http10
+    :want_service_https :needs_component_httpd
+{
+    my ($self) = @_;
+
+    # skip past setup logs
+    $self->{instance}->getsyslog();
+
+    my $https = $self->{instance}->get_service('https');
+    my $alpn_map = [ 'http/1.0' ];
+
+    my $response = $self->do_https_request($https, $alpn_map);
+
+    $self->assert_num_equals(1, $response->{success});
+    # XXX if we negotiated HTTP/1.0 via ALPN, the server still talks HTTP/1.1
+    # XXX anyway...
+    $self->assert_str_equals('HTTP/1.1', $response->{protocol});
+    $self->assert_alpn_protocol(undef, $alpn_map->[0]);
```

Comment:
@ksmurchison Ignoring ALPN for a minute, how do we handle clients that only support HTTP/1.0?  I'm thinking about if we can extend the `do_h2` stuff after the `starttls()` call to also set up HTTP/1.0 vs HTTP/1.1 based on the ALPN negotiation.

Reply:
Looks like that's handled in `txn->flags.ver`, but that isn't available until cmdloop, hmm.

While I'm poking around, I notice we also call `starttls()` if (I think) the client requests to upgrade the connection to one using TLS.  I guess this is the http equivalent of imap STARTTLS.  Perhaps we need to check `tls_get_alpn_protocol()` there too.

Reply:
@elliefm With HTTP/1.x, the client determines the protocol used on the request line.  If the client uses HTTP/1.0, the server will respond as such.  If the client negotiates 1.0 via ALPN and uses 1.1 or vice-versa, I don't see that as a huge problem.  The major difference is that HTTP/1.0 has non-persistent connections by default.

As far as upgrading from plaintext to TLS, this code is only used in a Murder.  No existing HTTP clients that I'm aware of use the Upgrade mechanism for anything other than bootstrapping websockets.  Even upgrading from HTTP/1.1 to h2 via Upgrade has been deprecated.  I wouldn't worry about ALPN here.

Reply:
Ahh okay, I won't worry about it then.  Thanks!